community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValuePath.java

```diff
@@ -19,127 +19,441 @@
+ *      Rel ID :     1     1     2     3
+ * </pre>
+ * And negating relationships that are traversed backwards, we get:
+ * <pre>
+ *               (a)--&gt;(b)&lt;--(a)--&gt;(c)--&gt;(d)
+ *     Node ID :  0     1     0     2     3
+ *      Rel ID :    +1    -1    +2    +3
+ * </pre>
+ * Finally, we can drop the leading 0 from the Path structure and combine the
+ * sequence information with the other path data. This results in an overall
+ * Path representation as below:
+ * <pre>
+ *     Structure Path 'P' {
+ *         nodes         = [(a), (b), (c), (d)]
+ *         relationships = [(a:b), (a:c), (c:d)]
+ *         sequence      = [+1, 1, -1, 0, +2, 2, +3, 3]
+ *     }
+ * </pre>
+ * Therefore, for the original path containing five nodes and four relationships,
+ * we end up transmitting only four nodes, three unbound relationships and eight
+ * bytes of sequence data (small integers require only a single byte in PackStream).
+ */
 public class ValuePath implements Path
 {
-    private final PropertyContainer[] entities;
+    private static final int STRUCT_FIELD_COUNT = 5;
+
+    private static List<Node> packNodes( Neo4jPack.Packer packer, Path path )
+            throws IOException
+    {
+        // TODO: rewrite this without object creation
+        // TODO: (might require additions to the Path interface)
+        List<Node> nodes = new ArrayList<>( path.length() + 1 );
+        for ( Node node : path.nodes() )
+        {
+            if ( !nodes.contains( node ) )
+            {
+                nodes.add( node );
+            }
+        }
+        packer.packListHeader( nodes.size() );
+        for ( Node node : nodes )
+        {
+            ValueNode.pack( packer, node );
+        }
+        return nodes;
+    }
+
+    private static List<Relationship> packRelationships( Neo4jPack.Packer packer, Path path )
+            throws IOException
+    {
+        // TODO: rewrite this without object creation
+        // TODO: (might require additions to the Path interface)
+        List<Relationship> relationships = new ArrayList<>( path.length() );
+        for ( Relationship rel : path.relationships() )
+        {
+            if ( !relationships.contains( rel ) )
+            {
+                relationships.add( rel );
+            }
+        }
+        packer.packListHeader( relationships.size() );
+        for ( Relationship rel : relationships )
+        {
+            ValueUnboundRelationship.pack( packer, ValueUnboundRelationship.unbind( rel ) );
+        }
+        return relationships;
+    }
+
+    public static void pack( Neo4jPack.Packer packer, Path path ) throws IOException
+    {
+        packer.packStructHeader( 3, Neo4jPack.PATH );
+
+        List<Node> nodes = packNodes( packer, path );
+        List<Relationship> relationships = packRelationships( packer, path );
+
+        boolean firstNode = true;
+        Node node = null;
+        packer.packListHeader( 2 * path.length() );
```

Comment:
If we go ahead and look at the List -> Map optimization, we could also look at the inversion optimization we discussed here. So, rather than building up the Node/Rel lists ahead of time, we build them lazily as we iterate through the path.

Doing so might be easier if, rather than combining serialization into the Path value object, we created a stateful `PathFormat` class or something. Neo4jPacker would have a `PathFormat` instance it delegates to for serializing paths, and `PathFormat` could have re-usable Set instances as fields.